notifications_utils/logging/celery.py

```diff
@@ -0,0 +1,52 @@
+import os
+import warnings
+from logging.config import dictConfig
+
+from celery.signals import setup_logging
+
+
+def set_up_logging(logger):
+    if logger is None or not hasattr(logger, "warning"):
+        raise AttributeError("The provided logger object is invalid.")
+
+    def custom_showwarning(message, category, filename, lineno, file=None, line=None):
+        log_entry = {
+            "level": "WARNING",
+            "message": str(message),
+            "category": category.__name__,
+            "filename": filename,
+            "lineno": lineno,
+        }
+        logger.warning(str(message), log_entry)
+
+    warnings.showwarning = custom_showwarning
+
+
+@setup_logging.connect
```

Comment:
Even though it's the documented way of doing this, I don't like things implicitly doing things on import - i'd much rather instead we did `setup_logging.connect(setup_logging_connect)` from inside `set_up_logging`

Reply:
e.g. I can imagine someone getting confused, "but I commented out the set_up_logging call!"

Reply:
I agree. Unfortunately if I put the the:

```
@setup_logging.connect
def setup_logging_connect(*args, **kwargs):
```

inside the set_up_logging function it no longer works. I can only assume that celery is doing stuff on import.

You got any other ideas on how to do this that I might not have tried?

Reply:
I also agree, but I think it's just the way Celery works - if you don't register a listener on the setup_logging signal then Celery sets up its default logging automatically (presumably on import).

https://docs.celeryq.dev/en/stable/userguide/signals.html#setup-logging

Reply:
I don't feel that strongly about it, I think it's fine the way you've done it

(I suspect `after_setup_logger` won't work for the same reason: because you're registering the listener after the signal has already fired)

Reply:
If you leave `def setup_logging_connect(...` in the same place and instead of adding the decorator, add a line to `set_up_logging` that does `setup_logging.connect(setup_logging_connect)`, it should be near-functionally-identical to using it as a decorator (apart from it not having import-time side effects) given the way `celery_dummy.py` calls `set_up_logging` only a few lines after importing this module.

Reply:
The test is passing this way. Nice.